File: 1236-n-th-tribonacci-number/1236-n-th-tribonacci-number.cpp
class Solution {
public:
    int f(int index,vector<int>&dp){
        if(index==0){
            return 0;
        }
        if(index==1) return 1;
        if(index==2) return 1;
        if(dp[index]!=-1){
            return dp[index];
        }
        int one = f(index-1,dp);
        int two=f(index-2,dp);
        int three=f(index-3,dp);
        return dp[index]=one+two+three;
    }
    int tribonacci(int n) {
        vector<int>dp(n+1,-1);
        return f(n,dp);
    }
};

File: 1236-n-th-tribonacci-number/1236-n-th-tribonacci-number.java
class Solution {
    public int tribonacci(int n) {
    int[] dp = new int[n+1];
    Arrays.fill(dp,0);
    return func(n,dp);
    }
    private int func(int n ,int[] dp)
    {
        if(n==0)
            return 0;
        if (n<3)
            return 1;
        if(dp[n]!=0)
            return dp[n];
        dp[n] = func(n-1,dp)+func(n-2,dp)+func(n-3,dp);
        return dp[n];
    }
}